CI boilerplate reduction (#34272)
LLNL/spack: lib/spack/docs/pipelines.rst

CI Pipelines
============

Spack provides commands that support generating and running automated build pipelines in CI instances.  At the highest
level it works like this: provide a spack environment describing the set of packages you care about, and include a
description of how those packages should be mapped to Gitlab runners.  Spack can then generate a ``.gitlab-ci.yml``
file containing job descriptions for all your packages that can be run by a properly configured CI instance.  When
run, the generated pipeline will build and deploy binaries, and it can optionally report to a CDash instance
regarding the health of the builds as they evolve over time.

------------------------------
Getting started with pipelines
------------------------------

To get started with automated build pipelines a Gitlab instance with version ``>= 12.9``
(more about Gitlab CI `here <https://about.gitlab.com/product/continuous-integration/>`_)
with at least one `runner <https://docs.gitlab.com/runner/>`_ configured is required. This
can be done quickly by setting up a local Gitlab instance.

> It is possible to set up pipelines on gitlab.com the builds there are limited to

Reviewer: ```suggestion
It is possible to set up pipelines on gitlab.com, but the builds there are limited to
```

60 minutes and generic hardware.  It is possible to
`hook up <https://about.gitlab.com/blog/2018/04/24/getting-started-gitlab-ci-gcp>`_
Gitlab to Google Kubernetes Engine (`GKE <https://cloud.google.com/kubernetes-engine/>`_)
or Amazon Elastic Kubernetes Service (`EKS <https://aws.amazon.com/eks>`_), though those
topics are outside the scope of this document.

After setting up a Gitlab instance for running CI, the basic steps for setting up a build pipeline are as follows:

> #. Create a repository in the Gitlab instance CI and a runner enabled.

Reviewer: ```suggestion
#. Create a repository in the Gitlab instance CI and enable a runner.
```

Author: Since enabling CI and Runners are configs in each gitlab repo, I am trying to convey that the repo they create must have those components enabled. Would it be better as `with CI and a runner enabled`?

Reviewer: Yeah, that sounds good to me.

     mirrors: { "mirror": "s3://spack-public/mirror" }

     ci:
       enable-artifacts-buildcache: True
       rebuild-index: False
       pipeline-gen:
       - any-job:
           before_script:
             - git clone ${SPACK_REPO}
             - pushd spack && git checkout ${SPACK_CHECKOUT_VERSION} && popd
             - . "./spack/share/spack/setup-env.sh"
       - build-job:
           tags: [docker]
           image:
             name: ghcr.io/scottwittenburg/ecpe4s-ubuntu18.04-runner-x86_64:2020-09-01
             entrypoint: [""]

> 

Reviewer: Does it make sense below here, to mention why you don't see any `script` section, and that one from a config scope will be inserted for you by default?

a particular build locally.

------------------------------------
Job Types
------------------------------------

^^^
Rebuild (build)
^^^

> Rebuild jobs, denoted as ``build-job``'s in the ``pipeline-gen`` list, are jobs associated with a concrete spec that have been marked for rebuild. By default a simple script for doing rebuild is generated, but may be modified as needed.

Reviewer: ```suggestion
Rebuild jobs, denoted as ``build-job``'s in the ``pipeline-gen`` list, are jobs associated with concrete specs that have been marked for rebuild. By default a simple script for doing rebuild is generated, but may be modified as needed.
```

The default script does three main steps, change directories to the pipelines concrete environment, activate the concrete environment, and run the ``spack ci rebuild`` command:

.. code-block:: bash

>   cd ${concrete_environment_dir}

Reviewer: Adding a newline here fixes the doc build error for me locally.  I did find some mention of this buried in the `Directives` section of the sphinx rst docs [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#directives).

  spack env activate --without-view .
  spack ci rebuild

.. _rebuild_index:

> ^^^

Reviewer: You have to extend these to the end of the text they surround, or else they end up rendered like this:

![broken_section](https://user-images.githubusercontent.com/6527504/213294940-78d19e3f-2a7d-4f4a-b6a2-8db784a3f828.png)

To make sure the buildcache index is up to date at the end of your pipeline,
spack generates a job to update the buildcache index of the target mirror
at the end of each pipeline by default.  You can disable this behavior by
adding ``rebuild-index: False`` inside the ``ci`` section of your
spack environment.

> Reindex jobs do not allow modifying the ``script`` attribute since it is automatically generate using the target mirror listed in the ``mirrors::mirror`` configuration.

Reviewer: ```suggestion
Reindex jobs do not allow modifying the ``script`` attribute since it is automatically generated using the target mirror listed in the ``mirrors::mirror`` configuration.
```

^^^
Signing (signing)
^^^

> This job is run after all of the rebuild jobs are completed and is intended to be used to sign the package binaries built by a protected CI run. Signing jobs are generated only if a signing job ``script`` is specified and the spack CI job type is protected. Note, if an ``any-job`` section contains a script, this will not implicitly create a ``signing`` job, a signing job may only exist if it is explicitly specified in the configuration with a ``script`` attribute. Specifying a signing job without a script does not create a signing job and the job configuration attributes will be ignored. Signing jobs are always assigned the runner tags ``aws``, ``protected``, and ``notary``.

Reviewer: Everywhere else in this document and others, it seems newlines have been used to control line length.  We should probably do the same with all your additions here.

^^^
Cleanup (cleanup)
^^^

> When using ``temporary-storage-url-prefix`` the cleanup job will destroy the mirror created for the parent Gitlab pipeline job. Cleanup jobs do not allow modifying the script, but do expect that the spack command is in the path and require a ``before_script`` to be specified that sources the ``setup-env.sh`` script.

Reviewer: ```suggestion
When using ``temporary-storage-url-prefix`` the cleanup job will destroy the mirror created for the associated Gitlab pipeline. Cleanup jobs do not allow modifying the script, but do expect that the spack command is in the path and require a ``before_script`` to be specified that sources the ``setup-env.sh`` script.
```

.. code-block:: yaml

  spack:
     ci:
       pipeline-gen:
       - noop-job:
           tags: ['custom', 'tag']
           image:
             name: 'some.image.registry/custom-image:latest'
             entrypoint: ['/bin/bash']
           script::
             - echo "Custom message in a custom script"

The example above illustrates how you can provide the attributes used to run
the NO-OP job in the case of an empty pipeline.  The only field for the NO-OP
job that might be generated for you is ``script``, but that will only happen

> if you do not provide one yourself. Notice in this example the ``script`` uses the ``::`` notation to perscribe override behavior. Without this, the ``echo`` command would have been prepended to the automatically generated script rather than replacing it.

Reviewer: ```suggestion
if you do not provide one yourself. Notice in this example the ``script`` uses the ``::`` notation to prescribe override behavior. Without this, the ``echo`` command would have been prepended to the automatically generated script rather than replacing it.
```

------------------------------------
ci.yaml
------------------------------------

Here's an example of a spack configuration file describing a build pipeline:

.. code-block:: yaml

  ci:
    target: gitlab

    rebuild_index: True

    broken-specs-url: https://broken.specs.url

    broken-tests-packages:
    - gptune

    pipeline-gen:
    - submapping:
      - match:
          - os=ubuntu18.04
        build-job:
          tags:
            - spack-kube
          image: spack/ubuntu-bionic
      - match:
          - os=centos7
        build-job:
          tags:
            - spack-kube
          image: spack/centos7

    cdash:
      build-group: Release Testing
      url: https://cdash.spack.io
      project: Spack
      site: Spack AWS Gitlab Instance

The ``ci`` config section is used to configure how the pipeline workload should be
generated, mainly how the jobs for building specs should be assigned to the
configured runners on your instance. The main section for configuring pipelines
is ``pipeline-gen``, which is a list of job attribute sections that are merged,

> using the same rules as Spack configs (TODO:ref:`spack_configs`), from the bottom up.

Reviewer: What is the TODO about here?  Just making a link to another section of the docs, or something more?

Author: yeah, TODO, figure out how to reference the other section in the docs.

> The order sections are applied is to be consistent with how spack orders scope precendense when merging lists.

Reviewer: ```suggestion
The order sections are applied is to be consistent with how spack orders scope precedence when merging lists.
```

There are two main section types, ``<type>-job`` sections and ``submapping``
sections.

^^^
Job Attribute Sections
^^^

Each type of job may have attributes added or removed via sections in the ``pipeline-gen`` list. Job type specific attributes may be specified using the keys ``<type>-job`` to add attributes to all jobs of type ``<type>`` or ``<type>-job-remove`` to remove attributes of type ``<type>``. Each section may only contain one type of job attribute specification, ie. , ``build-job`` and ``noop-job`` may not coexist but ``build-job`` and ``build-job-remove`` may. Note, ``*-remove`` specifications are applied before the additive attribute specification.

> All of the attributes specified are forwarded to the generated CI jobs, however special treatment is applied to the attributes ``tags``, ``image``, ``variables``, ``script``, ``before_script``, and ``after_script`` as they are components recognized explicitly by the Spack CI generator. For the ``tags`` attribute, Spack will remove reserved tags (:ref:`reserved_tags`) from all jobs specified in the config. In some cases, such as for ``signing`` jobs, reserved tags will be added back based on the type of CI that is being run.

Reviewer: The next two paragraphs following this one probably need updated to remove mention of `runner-attributes`, which I don't think is used anymore, is it?

provided, within your ``$CI_PROJECT_DIR``), activates that environment for
you, and invokes ``spack ci rebuild``.

> Sections that specify scripts (``script``, ``before_script``, ``after_script``) are all read as lists of commands or lists of lists of commands. It is recommended to write scripts as lists of lists if scripts will be composes via merging. The default behavior of merging lists will remove duplicate commands and potentially apply unwanted reordering, where as merging lists of lists will preserve the local ordering and never removes duplicate commands. When writting commands to the CI target script, all lists are expanded and flattened into a single list.

Reviewer: ```suggestion
Sections that specify scripts (``script``, ``before_script``, ``after_script``) are all read as lists of commands or lists of lists of commands. It is recommended to write scripts as lists of lists if scripts will be composed via merging. The default behavior of merging lists will remove duplicate commands and potentially apply unwanted reordering, whereas merging lists of lists will preserve the local ordering and never removes duplicate commands. When writing commands to the CI target script, all lists are expanded and flattened into a single list.
```

^^^
Submapping Sections
^^^

> A special case of attribute specification is the ``submapping`` section which may be used to apply job attributes to build jobs based on the package spec associated with the rebuild job. Submapping is specified as a list of spec ``match`` lists associated with ``build-job``/``build-job-remove`` sections. There are two option for ``match_behavior``, either ``first`` or ``merge`` may be specified. In either case, the ``submapping`` list is processed from the bottom up, and then each ``match`` lists is searched for a string that satisfies the check ``spec.satisfies({match_item})`` for each concrete spec.

Reviewer: ```suggestion
A special case of attribute specification is the ``submapping`` section which may be used to apply job attributes to build jobs based on the package spec associated with the rebuild job. Submapping is specified as a list of spec ``match`` lists associated with ``build-job``/``build-job-remove`` sections. There are two options for ``match_behavior``, either ``first`` or ``merge`` may be specified. In either case, the ``submapping`` list is processed from the bottom up, and then each ``match`` list is searched for a string that satisfies the check ``spec.satisfies({match_item})`` for each concrete spec.
```

^^^^^
CDash
^^^^^

The optional ``cdash`` section provides information that will be used by the
``spack ci generate`` command (invoked by ``spack ci start``) for reporting
to CDash.  All the jobs generated from this environment will belong to a
"build group" within CDash that can be tracked over time.  As the release
progresses, this build group may have jobs added or removed. The url, project,
and site are used to specify the CDash instance to which build results should
be reported.

Take a look at the
`schema <https://github.com/spack/spack/blob/develop/lib/spack/spack/schema/ci.py>`_
for the gitlab-ci section of the spack environment file, to see precisely what
syntax is allowed there.

.. _reserved_tags:

^^^^^^^^^^^^^
Reserved Tags
^^^^^^^^^^^^^

Spack has a subset of tags (``public``, ``protected``, and ``notary``) that it reserves
for classifying runners that may require special permissions or access. The tags
``public`` and ``protected`` are used to distinguish between runners that use public
permissions and runners with protected permissions. The ``notary`` tag is a special tag
that is used to indicate runners that have access to the highly protected information

> used for signing binaries using a the ``signing`` job.

Reviewer: ```suggestion
used for signing binaries using the ``signing`` job.
```
